### PR TITLE
Remove 2D mon_at, npc_at, critter_at, is_empty etc.

### PR DESCRIPTION
**Pull request by Coolthulhu**
Includes #12307

Here I remove 2D overloads for `game` functions that find creatures or check sunlight, shelter or free space.

Had to update some functions like:
- Knockback
- Fungus spawning monattacks
- Marlossification of area
- Computers summoning manhacks or secubots

Also changed some scent code. Fixed a bug that caused the scent not to propagate while the player is standing still (it should stop propagating after 1000 turns, but did so after 1).

Discovered a bug in scent code: it uses two different boundary calculations, which results in "pockets" of scent that doesn't decay. Also, player moving back and forth will re-create scent that was "deleted" due to player moving away. Didn't fix them here, though. Those are very minor and can wait for 3D scent update.

### CODE REVIEW

**Comment by DavidKeaton** [comment]
that was the odd pockets thing from a while back

File: src/game.cpp [hunk 1]
```diff
@@ -2012,12 +2013,11 @@ input_context game::get_player_input(std::string &action)
...
 
-                    for (int i = 0; i < (int)iter->getText().length(); ++i) {
-                        const int dex = mon_at(iter->getPosX() + i, iter->getPosY());
+                    tripoint tmp( iter->getPosX(), iter->getPosY(), get_levz() );
+                    for( tmp.x = iter->getPosX(); tmp.x < iter->getPosX() + (int)iter->getText().length(); ++tmp.x ) {
+                        const int dex = mon_at( tmp );
 
                         if (dex != -1 && u.sees(zombie(dex))) {
-                            i = -1;
-
```
Comment (BevapDin) [inline src/game.cpp]: This looks buggy. The inner loop calls `advanceStepOffset`, which changes `iStepOffset`, which changes the result of `getPosX` and `getPosY`. Because the result of `getPosY` is only used once to initialize `tmp`, the change here won't affect the loop.

`i = -1` is  probably used to restart the loop (the location of the on-screen-text has changed), this behavior is completely gone in the code. (The loop checks whether the on-screen-text would override a monster glyph and (if so) moves the text a bit, but it may again override another monster glyph at the new location.)

In short: it's probably better to create a temporary only for the calling of `mon_at`. Btw. the code ignores NPCs as if they didn't exist (it should perhaps use `critter_at`).

File: src/game.cpp [hunk 2]
```diff
@@ -3000,33 +3000,33 @@ bool game::handle_action()
...
-        return nulscent; // Out-of-bounds - null scent
-    }
-    return grscent[x][y];
+// TODO: Unify this and scent area used in update_scent to fix "scent pocket" bug
+bool outside_scent_radius( const tripoint &p ) {
+    return p.x < (SEEX * MAPSIZE / 2) - SCENT_RADIUS || p.x >= (SEEX * MAPSIZE / 2) + SCENT_RADIUS ||
+           p.y < (SEEY * MAPSIZE / 2) - SCENT_RADIUS || p.y >= (SEEY * MAPSIZE / 2) + SCENT_RADIUS;
 }
 
 int &game::scent( const tripoint &p ) // A wrapper for now
```
Comment (BevapDin) [inline src/game.cpp]: Comment needs to be removed or updated or whatever, it makes no sense currently.

Comment (Coolthulhu) [inline src/game.cpp]: The comment refers to the fact that area in `outside_scent_radius` is different from the one in `update_scent`.

`update_scent` has an area like `u.pos() +/- SCENT_RADIUS`, while `outside_scent_radius` has a hardcoded dependence on the center of the map. This discrepancy results in "pockets" of scent that do not decay.
I didn't fix this one here, because centering the scent map on the center of the bubble could result in jarring changes, while centering around the player would slow down scent processing.

File: src/game.cpp [hunk 3]
```diff
@@ -12298,13 +12240,8 @@ bool game::plmove(int dx, int dy)
...
         // tile is impassable
-        while ((m.move_cost(x + tunneldist * (x - u.posx()), y + tunneldist * (y - u.posy())) == 0) ||
-               // a monster is there
-               ((mon_at(x + tunneldist * (x - u.posx()), y + tunneldist * (y - u.posy())) != -1 ||
-                 // so keep tunneling
-                 npc_at(x + tunneldist * (x - u.posx()), y + tunneldist * (y - u.posy())) != -1) &&
-                // assuming we've already started
-                tunneldist > 0)) {
+        tripoint dest( x + tunneldist * (x - u.posx()), y + tunneldist * (y - u.posy()), u.posz() );
+        while( ( m.move_cost(dest) == 0 || mon_at(dest) != -1 || npc_at(dest) != -1 ) && tunneldist > 0 ) {
```
Comment (BevapDin) [inline src/game.cpp]: Hello infinite loop, my old friend. This condition only depends on `dest` and `tunneldist` (which is incresed in the loop body, so once it `tunneldist > 0` becomes true, it stays true, until it overflows). But `dest` is not changed in the loop body.

It's not actually infinite as the loop is left when tunneldist reaches 25 or when the character has not enough power left.

File: src/game.cpp (hunk 2 above)
Comment (BevapDin) [inline src/game.cpp]: I actually meant the `// A wrapper for now` - is it still a wrapper? It looks like it is the actual, real function to be used. If it's still a wrapper, what does it wrap?

**Comment by BevapDin** [comment]
Aside from the mentioned things above, this is fine.

----

Some notes through:

Inconsistency: Sometimes only a temporary wrapper is added: `mon_at(x,y)` becomes `mon_at(tripoint(x,y,defaulZ))`, other times a new tripoint variable is introduced and used (even if it is (currently) only used once).

Some calls (e.g. to `veh_at` in editmap.cpp) have been changed to tripoints, some haven't (even through other functions on the very same line now use tripoints). E.g.
```C++
if( veh_at(x, y) && mon_at(x, y))
// has been changed to:
tripoint p(x, y, defaulZ);
if( veh_at(x, y) && mon_at(p))
// Note the remaining usage of x and y
```

Sometimes a bigger block of code changes to use tripoints, e.g. in the diff of event.cpp: the code that spawns mon_dark_wyrm is changed differently than the code that spawns mon_amigara_horror.

----

Also code that loops over a 2D area using a tripoint is a bit inconsistent (the iteration only touches the x,y components, so z is initialized upfront):
````C++
tripoint tmp;
tmp.z = posz();
// vs.
tripoint tmp = pos();
// vs.
tripoint dest( 0, 0, posz() );
```
Either are correct (initializing z) and fine (x,y are overridden in the loop anyway, so their initial value does not matter).

Similar:
```
-            int x, y;
+            tripoint dest( 0, 0, posz() );
             int tries = 0;
             do {
-                x = posx() + rng(-4, 4);
-                y = posy() + rng(-4, 4);
+                dest.x = posx() + rng(-4, 4);
+                dest.y = posy() + rng(-4, 4);
```
vs
```
-                int x, y;
+                tripoint dest( 0, 0, posz() );
+                int &x = dest.x;
+                int &y = dest.y;
```
(Personally I like the `int &x` solution, at least as long `x` and `y` are still heavily referenced through the following code. It assure correct behavior whether `x,y` is or `dest` is used.)

----

(It's a bit late for this, my fault.)
It would be easier to review if you changed one function (e.g. `mdeath::ratking`) to use only tripoints (no calls that only include x,y) - add *simple* wrapper function to wrap the calls (e.g. if it calls `is_empty` which currently only supports x,y parameters, add `is_empty` overload that uses tripoints, but don't change any other calls to `is_empty` outside of `mdeath::ratking`).

This would also allow easy testing of that function.

**Comment by Coolthulhu** [comment]
Most of the inconsistency results from me wanting to "secure" the 3D code. As new code is added, it often uses the legacy 2D code. This can result in conflicts and extra work.

As for the x/y vs tripoint inconsistency: I'll change it all to tripoints later. Currently the x/y code is there only because otherwise I'd have to rewrite the whole function, possibly causing issues like with that tunneling.

> Also code that loops over a 2D area using a tripoint is a bit inconsistent (the iteration only touches the x,y components, so z is initialized upfront)

I usually do the `int &x = p.x`, but for some loops over a tiny area (-1 to 1) this is a bit too verbose. In those I address the members directly.
But if you think it is a better idea to have the same kind of code everywhere, I can always use the references.

File: src/game.cpp [hunk 4]
```diff
@@ -12296,15 +12239,10 @@ bool game::plmove(int dx, int dy)
...
-               ((mon_at(x + tunneldist * (x - u.posx()), y + tunneldist * (y - u.posy())) != -1 ||
-                 // so keep tunneling
-                 npc_at(x + tunneldist * (x - u.posx()), y + tunneldist * (y - u.posy())) != -1) &&
-                // assuming we've already started
-                tunneldist > 0)) {
+        int tunneldist = 0;
+        do {
             //add 1 to tunnel distance for each impassable tile in the line
             tunneldist += 1;
             if (tunneldist * 250 > u.power_level) { //oops, not enough energy! Tunneling costs 250 bionic power per impassable tile
```
Comment (BevapDin) [inline src/game.cpp]: If I have exact 500 power, I should be able to tunnel 2 squares, this test is now wrong. (The previous check was for the location that was impassable, now it's for a location that *might* be impassable or it might be the end point where the character appears.)

Moving the condition to the end also changes how `dest` is derived from `tunneldist`. The very first iteration checks for `dest.x == x + 1 * (x - u.posx()) == x + x - u.posx()`, assuming that `x == posx + 1` (attempt to move north or whatever the x-axis is), the check is for `dest.x == (posx + 1) + (posx + 1) - posx == posx + 2`, the place at `posx+1` is never checked at all.

**Comment by BevapDin** [comment]
> I usually do the  int &x = p.x , but for some loops over a tiny area (-1 to 1) this is a bit too verbose. In those I address the members directly.

Well, that is a form of consistency.

Ideally (I'm allowed to dream) the iteration is replaced by some generic thing, e.g. some special iterator classes that make up the points during the iteration.

```C++
// Using a generator function that generates the points on one call:
for( const tripoint & p : closest_points_first( center, radius ) ) {
}

// or better with a pseudo-container, iterates over (x-r,y-r)...(x+r,y+r):
class iterator_over_flat_rectangle {
    tripoint center;
    int radius;
    iterator begin() const;
    iterator end() const;
};
// And all the magic basically happens inside of iterator::operator++ and iterator::operator*
// They make up the point during the iteration, using a fixed amount of memory (no dynamic container),
// and don't generate more points than needed when the loop is left early.
for( const tripoint & p : iterator_over_rectangle( center, radius ) ) {
    if( is_empty(p)) ...
}
// Same as the other class, but iterates over different z-level as well
class iterator_over_rectangle_z {
    ...
};
```

> But if you think it is a better idea to have the same kind of code everywhere, I can always use the references.

It's not a problem as long as the code is correct and it's not a big difference anyway. If you have specific reason to choose one way over the other, that's fine. Only if there is no reason to use a specific style, stick with one and the same.

**Comment by Coolthulhu** [comment]
> Ideally (I'm allowed to dream) the iteration is replaced by some generic thing, e.g. some special iterator classes that make up the points during the iteration.

Didn't think of it, I kept thinking of the `yield`, which C++ doesn't have. Sounds really good, I may try implementing it soon.
This could also be extended to tile accesses (for example, with the `maptile` wrapper) to speed up functions operating on area (such as NPCs searching for items).
